File: InfoCenter/src/main/java/com/fallahpoor/infocenter/fragments/battery/BatteryViewModel.java
package com.fallahpoor.infocenter.fragments.battery;

import android.app.Application;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.BatteryManager;

import com.fallahpoor.infocenter.R;
import com.fallahpoor.infocenter.Utils;

import java.text.NumberFormat;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

public class BatteryViewModel extends AndroidViewModel {

    private Utils utils;
    private NumberFormat percentFormat;
    private Context context;
    private MutableLiveData<BatteryState> batteryStateMutableLiveData;
    private BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            updateBatteryStateLiveData(intent);
        }
    };

    private void updateBatteryStateLiveData(Intent intent) {

        int status;
        int level;
        int health;
        int plugged;
        int temperature;
        int voltage;
        String technology;

        status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, 0);
        level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, 0);
        health = intent.getIntExtra(BatteryManager.EXTRA_HEALTH, 0);
        plugged = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, 0);
        technology = intent.getStringExtra(BatteryManager.EXTRA_TECHNOLOGY);
        temperature = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0) / 10;
        voltage = intent.getIntExtra(BatteryManager.EXTRA_VOLTAGE, 0);

        BatteryState batteryState = new BatteryState();

        batteryState.setStatus(getStatus(status));
        batteryState.setLevel(getLevel(level));
        batteryState.setHealth(getHealth(health));
        batteryState.setPlugged(getPlugged(status, plugged));
        batteryState.setTechnology(getTechnology(technology));
        batteryState.setTemperature(getTemperature(temperature));
        batteryState.setVoltage(getVoltage(voltage));

        batteryStateMutableLiveData.setValue(batteryState);

    }

    public BatteryViewModel(@NonNull Application application) {
        super(application);
        context = application;
        batteryStateMutableLiveData = new MutableLiveData<>();
        utils = new Utils(context);
        percentFormat = NumberFormat.getPercentInstance(utils.getLocale());
    }

    public LiveData<BatteryState> getBatteryStateLiveData() {
        context.registerReceiver(broadcastReceiver, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
        return batteryStateMutableLiveData;
    }

    private String getStatus(int intStatus) {

        String status;

        if (intStatus == BatteryManager.BATTERY_STATUS_CHARGING
                || intStatus == BatteryManager.BATTERY_STATUS_FULL) {
            status = context.getString(R.string.bat_sub_item_charging);
        } else {
            status = context.getString(R.string.bat_sub_item_discharging);
        }

        return status;

    }

    private String getLevel(int level) {
        return percentFormat.format((double) level / 100);
    }

    private String getHealth(int intHealth) {

        String health;

        switch (intHealth) {
            case BatteryManager.BATTERY_HEALTH_COLD:
            case BatteryManager.BATTERY_HEALTH_GOOD:
                health = context.getString(R.string.bat_sub_item_good);
                break;
            case BatteryManager.BATTERY_HEALTH_DEAD:
                health = context.getString(R.string.bat_sub_item_dead);
                break;
            case BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE:
                health = context.getString(R.string.bat_sub_item_over_voltage);
                break;
            case BatteryManager.BATTERY_HEALTH_OVERHEAT:
                health = context.getString(R.string.bat_sub_item_overheat);
                break;
            default:
                health = context.getString(R.string.unknown);
                break;
        }

        return health;

    }

    private String getPlugged(int intStatus, int intPlugged) {

        String plugged;

        if (intStatus == BatteryManager.BATTERY_STATUS_CHARGING
                || intStatus == BatteryManager.BATTERY_STATUS_FULL) {
            switch (intPlugged) {
                case BatteryManager.BATTERY_PLUGGED_AC:
                    plugged = context.getString(R.string.bat_sub_item_yes_ac);
                    break;
                case BatteryManager.BATTERY_PLUGGED_USB:
                    plugged = context.getString(R.string.bat_sub_item_yes_usb);
                    break;
                case BatteryManager.BATTERY_PLUGGED_WIRELESS:
                    plugged = context.getString(R.string.bat_sub_item_yes_wireless);
                    break;
                default:
                    plugged = context.getString(R.string.unknown);
            }
        } else {
            plugged = context.getString(R.string.no);
        }

        return plugged;

    }

    private String getTechnology(String technology) {

        String tech;

        switch (technology.toLowerCase(Locale.US)) {
            case "li-ion":
                tech = context.getString(R.string.bat_sub_item_lithium_ion);
                break;
            case "li-poly":
                tech = context.getString(R.string.bat_sub_item_lithium_polymer);
                break;
            default:
                tech = technology;
        }

        return tech;

    }

    private String getTemperature(int temperature) {
        return String.format(utils.getLocale(), "%d %s", temperature,
                context.getString(R.string.bat_sub_item_centigrade));
    }

    private String getVoltage(int voltage) {
        return String.format(utils.getLocale(), "%d %s", voltage,
                context.getString(R.string.bat_sub_item_milli_volt));
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        context.unregisterReceiver(broadcastReceiver);
    }

}

File: InfoCenter/src/main/java/com/fallahpoor/infocenter/fragments/battery/BatteryViewModel.kt
package com.fallahpoor.infocenter.fragments.battery

import android.app.Application
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.fallahpoor.infocenter.R
import com.fallahpoor.infocenter.Utils
import java.text.NumberFormat
import java.util.*

class BatteryViewModel(application: Application) : AndroidViewModel(application) {

    private val utils: Utils
    private val percentFormat: NumberFormat
    private val context: Context
    private val batteryStateMutableLiveData: MutableLiveData<BatteryState>
    private val broadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            updateBatteryStateLiveData(intent)
        }
    }

    val batteryStateLiveData: LiveData<BatteryState>
        get() {
            context.registerReceiver(broadcastReceiver, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
            return batteryStateMutableLiveData
        }

    private fun updateBatteryStateLiveData(intent: Intent) {

        val status: Int = intent.getIntExtra(BatteryManager.EXTRA_STATUS, 0)
        val level: Int = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, 0)
        val health: Int = intent.getIntExtra(BatteryManager.EXTRA_HEALTH, 0)
        val plugged: Int = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, 0)
        val temperature: Int = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0) / 10
        val voltage: Int = intent.getIntExtra(BatteryManager.EXTRA_VOLTAGE, 0)
        val technology: String = intent.getStringExtra(BatteryManager.EXTRA_TECHNOLOGY)

        val batteryState = BatteryState(
            getStatus(status),
            getLevel(level),
            getHealth(health),
            getPlugged(status, plugged),
            getTechnology(technology),
            getTemperature(temperature),
            getVoltage(voltage)
        )

        batteryStateMutableLiveData.value = batteryState

    }

    init {
        context = application
        batteryStateMutableLiveData = MutableLiveData()
        utils = Utils(context)
        percentFormat = NumberFormat.getPercentInstance(utils.locale)
    }

    private fun getStatus(intStatus: Int): String {

        return if (intStatus == BatteryManager.BATTERY_STATUS_CHARGING ||
            intStatus == BatteryManager.BATTERY_STATUS_FULL
        ) {
            context.getString(R.string.bat_sub_item_charging)
        } else {
            context.getString(R.string.bat_sub_item_discharging)
        }

    }

    private fun getLevel(level: Int): String {
        return percentFormat.format(level.toDouble() / 100)
    }

    private fun getHealth(intHealth: Int): String {
        return when (intHealth) {
            BatteryManager.BATTERY_HEALTH_COLD,
            BatteryManager.BATTERY_HEALTH_GOOD -> context.getString(R.string.bat_sub_item_good)
            BatteryManager.BATTERY_HEALTH_DEAD -> context.getString(R.string.bat_sub_item_dead)
            BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE -> context.getString(R.string.bat_sub_item_over_voltage)
            BatteryManager.BATTERY_HEALTH_OVERHEAT -> context.getString(R.string.bat_sub_item_overheat)
            else -> context.getString(R.string.unknown)
        }
    }

    private fun getPlugged(status: Int, plugged: Int): String {

        return if (status == BatteryManager.BATTERY_STATUS_CHARGING ||
            status == BatteryManager.BATTERY_STATUS_FULL
        ) {
            when (plugged) {
                BatteryManager.BATTERY_PLUGGED_AC -> context.getString(R.string.bat_sub_item_yes_ac)
                BatteryManager.BATTERY_PLUGGED_USB -> context.getString(R.string.bat_sub_item_yes_usb)
                BatteryManager.BATTERY_PLUGGED_WIRELESS -> context.getString(R.string.bat_sub_item_yes_wireless)
                else -> context.getString(R.string.unknown)
            }
        } else {
            context.getString(R.string.no)
        }

    }

    private fun getTechnology(technology: String): String {
        return when (technology.toLowerCase(Locale.US)) {
            "li-ion" -> context.getString(R.string.bat_sub_item_lithium_ion)
            "li-poly" -> context.getString(R.string.bat_sub_item_lithium_polymer)
            else -> technology
        }
    }

    private fun getTemperature(temperature: Int): String {
        return String.format(
            utils.locale, "%d %s", temperature,
            context.getString(R.string.bat_sub_item_centigrade)
        )
    }

    private fun getVoltage(voltage: Int): String {
        return String.format(
            utils.locale, "%d %s", voltage,
            context.getString(R.string.bat_sub_item_milli_volt)
        )
    }

    override fun onCleared() {
        super.onCleared()
        context.unregisterReceiver(broadcastReceiver)
    }

}
